## Imports

In [1]:
!python -V

Python 3.10.12


In [2]:
import os
import sys
import json
import requests

In [3]:
import numpy as np

import tensorflow as tf
from tensorflow import keras

tf.__version__

2024-12-09 11:01:29.787541: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1733738489.976649   73511 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1733738490.016360   73511 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-12-09 11:01:30.426065: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


'2.18.0'

In [4]:
from tensorflow.keras.preprocessing.image import load_img
from tensorflow.keras.applications.xception import preprocess_input

In [5]:
import tensorflow.lite as tflite

In [6]:
!nvidia-smi

Mon Dec  9 11:01:34 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.127.08             Driver Version: 550.127.08     CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce GTX 1050        Off |   00000000:01:00.0 Off |                  N/A |
| N/A   41C    P8             N/A / ERR!  |       7MiB /   4096MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

## Homework

In this homework, we'll deploy the Straight vs Curly Hair Type model we trained in the 
[previous homework](../08-deep-learning/homework.md).

Download the model from here: 

https://github.com/alexeygrigorev/large-datasets/releases/download/hairstyle/model_2024_hairstyle.keras

In [7]:
!wget https://github.com/alexeygrigorev/large-datasets/releases/download/hairstyle/model_2024_hairstyle.keras

--2024-12-09 11:02:13--  https://github.com/alexeygrigorev/large-datasets/releases/download/hairstyle/model_2024_hairstyle.keras
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/426348925/df5735c1-9082-4b67-968e-866f268793f8?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241209%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241209T100214Z&X-Amz-Expires=300&X-Amz-Signature=530336f7aef49ae0af18668f4ae9570829d5f2af3629ecbac3ec764ab7295c2f&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dmodel_2024_hairstyle.keras&response-content-type=application%2Foctet-stream [following]
--2024-12-09 11:02:14--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/426348925/df5735c1-9082-4b67-968e-866f268793f8?X-Amz-

In [8]:
!ls -lh
!file -v
!file model_2024_hairstyle.keras
!unzip model_2024_hairstyle.keras

total 154M
-rw-rw-r-- 1 aztleclan aztleclan  289 dic  9 11:00 Dockerfile
-rw-rw-r-- 1 aztleclan aztleclan  53K dic  9 11:00 Homework.ipynb
-rw-rw-r-- 1 aztleclan aztleclan 4,9K dic  8 11:18 homework.md
-rw-rw-r-- 1 aztleclan aztleclan 1,5K dic  9 11:00 lambda_function.py
-rw-rw-r-- 1 aztleclan aztleclan 154M dic  2 15:28 model_2024_hairstyle.keras
file-5.41
magic file from /etc/magic:/usr/share/misc/magic
model_2024_hairstyle.keras: Zip archive data, at least v2.0 to extract, compression method=store
Archive:  model_2024_hairstyle.keras
 extracting: metadata.json           
 extracting: config.json             
 extracting: model.weights.h5        


In [9]:
#!rm model_2024_hairstyle.keras
!ls

config.json  Homework.ipynb  lambda_function.py  model_2024_hairstyle.keras
Dockerfile   homework.md     metadata.json	 model.weights.h5


In [10]:
model_path = 'model_2024_hairstyle.keras'

In [11]:
model = tf.keras.models.load_model(model_path)

# Verifica la arquitectura del modelo cargado
model.summary()

I0000 00:00:1733738576.463412   73511 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 3482 MB memory:  -> device: 0, name: NVIDIA GeForce GTX 1050, pci bus id: 0000:01:00.0, compute capability: 6.1


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 198, 198, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 99, 99, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 313632)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 64)             │    20,072,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 40,146,948 (153.15 MB)

 Trainable params: 20,073,473 (76.57 MB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 20,073,475 (76.57 MB)

## Question 1

Now convert this model from Keras to TF-Lite format.

What's the size of the **converted** model?

* 27 Mb
* 43 Mb
* 77 Mb
* 127 Mb

In [12]:
tflite_model_path = 'model_2024_hairstyle.tflite'

In [13]:
converter = tf.lite.TFLiteConverter.from_keras_model(model)

tflite_model = converter.convert()

with open(tflite_model_path, 'wb') as f_out:
    f_out.write(tflite_model)

INFO:tensorflow:Assets written to: /tmp/tmp6i7d6djc/assets


INFO:tensorflow:Assets written to: /tmp/tmp6i7d6djc/assets


Saved artifact at '/tmp/tmp6i7d6djc'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): TensorSpec(shape=(None, 200, 200, 3), dtype=tf.float32, name='input_layer')
Output Type:
  TensorSpec(shape=(None, 1), dtype=tf.float32, name=None)
Captures:
  139951173008400: TensorSpec(shape=(), dtype=tf.resource, name=None)
  139951173259280: TensorSpec(shape=(), dtype=tf.resource, name=None)
  139951173264032: TensorSpec(shape=(), dtype=tf.resource, name=None)
  139951173537984: TensorSpec(shape=(), dtype=tf.resource, name=None)
  139951173542560: TensorSpec(shape=(), dtype=tf.resource, name=None)
  139951173534640: TensorSpec(shape=(), dtype=tf.resource, name=None)


W0000 00:00:1733738605.439857   73511 tf_tfl_flatbuffer_helpers.cc:365] Ignored output_format.
W0000 00:00:1733738605.439885   73511 tf_tfl_flatbuffer_helpers.cc:368] Ignored drop_control_dependency.
2024-12-09 11:03:25.440596: I tensorflow/cc/saved_model/reader.cc:83] Reading SavedModel from: /tmp/tmp6i7d6djc
2024-12-09 11:03:25.441047: I tensorflow/cc/saved_model/reader.cc:52] Reading meta graph with tags { serve }
2024-12-09 11:03:25.441058: I tensorflow/cc/saved_model/reader.cc:147] Reading SavedModel debug info (if present) from: /tmp/tmp6i7d6djc
I0000 00:00:1733738605.445109   73511 mlir_graph_optimization_pass.cc:401] MLIR V1 optimization pass is not enabled
2024-12-09 11:03:25.445739: I tensorflow/cc/saved_model/loader.cc:236] Restoring SavedModel bundle.
2024-12-09 11:03:25.671743: I tensorflow/cc/saved_model/loader.cc:220] Running initialization op on SavedModel bundle at path: /tmp/tmp6i7d6djc
2024-12-09 11:03:25.678883: I tensorflow/cc/saved_model/loader.cc:466] SavedModel 

In [17]:
!ls -lh model_2024_hairstyle.tflite

-rw-rw-r-- 1 aztleclan aztleclan 77M dic  9 11:03 model_2024_hairstyle.tflite


What's the size of the **converted** model?

* 27 Mb
* 43 Mb
* **77 Mb**
* 127 Mb

## Question 2

To be able to use this model, we need to know the index of the input and 
the index of the output. 

What's the output index for this model?

* 3
* 7
* 13
* 24

In [18]:
interpreter = tflite.Interpreter(model_path=tflite_model_path)
interpreter.allocate_tensors()

input_index = interpreter.get_input_details()[0]['index']
output_index = interpreter.get_output_details()[0]['index']

INFO: Created TensorFlow Lite XNNPACK delegate for CPU.


In [19]:
input_index, output_index

(0, 13)

What's the output index for this model?

* 3
* 7
* **13**
* 24

## Preparing the image

You'll need some code for downloading and resizing images. You can use 
this code:

```python
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img
```

For that, you'll need to have `pillow` installed:

```bash
pip install pillow
```

Let's download and resize this image: 

https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg

Based on the previous homework, what should be the target size for the image?

In [20]:
from io import BytesIO
from urllib import request

from PIL import Image

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img


def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

In [21]:
image_url = "https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg"

In [22]:
ori_img = download_image(image_url)

In [23]:
ori_img.size

(1024, 1024)

In [24]:
target_size = (200, 200)

In [25]:
prepra_img = prepare_image(ori_img, target_size)

In [26]:
prepra_img.size

(200, 200)

## Question 3

Now we need to turn the image into numpy array and pre-process it. 

> Tip: Check the previous homework. What was the pre-processing 
> we did there?

After the pre-processing, what's the value in the first pixel, the R channel?

* 0.24
* 0.44
* 0.64
* 0.84


In [30]:
print(prepra_img.size)

#image_array = image_array.astype(np.float32)
image_array = np.array(prepra_img, dtype='float32')
print(image_array.shape)

# Add channel extra
image_array = np.expand_dims(image_array, axis=0) 
print(image_array.shape)

# Imagen to NumPy and Normalize
image_array = image_array / 255.0  # Normalize to [0, 1]
print(image_array.shape)

(200, 200)
(200, 200, 3)
(1, 200, 200, 3)
(1, 200, 200, 3)


In [31]:
first_pixel_r = image_array[0, 0, 0]  # El primer píxel (0,0) y el canal R (0)

In [32]:
first_pixel_r[0]

0.23921569

After the pre-processing, what's the value in the first pixel, the R channel?

* **0.24**
* 0.44
* 0.64
* 0.84


## Question 4

Now let's apply this model to this image. What's the output of the model?

* 0.293
* 0.493
* 0.693
* 0.893

In [33]:
interpreter.set_tensor(input_index, image_array)
interpreter.invoke()
preds = interpreter.get_tensor(output_index)

In [34]:
preds[0][0]

0.8937741

Now let's apply this model to this image. What's the output of the model?

* 0.293
* 0.493
* 0.693
* **0.893**

## Prepare the lambda code 

Now you need to copy all the code into a separate python file. You will 
need to use this file for the next two questions.

Tip: you can test this file locally with `ipython` or Jupyter Notebook 
by importing the file and invoking the function from this file.  


In [35]:
%%writefile lambda_function.py
import tflite_runtime.interpreter as tflite
from keras_image_helper import create_preprocessor
from io import BytesIO
from urllib import request
from PIL import Image
import numpy as np

def download_image(url):
    with request.urlopen(url) as resp:
        buffer = resp.read()
    stream = BytesIO(buffer)
    img = Image.open(stream)
    return img

def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img

tflite_model_path = 'model_2024_hairstyle_v2.tflite'
target_size = (200, 200)

interpreter = tflite.Interpreter(model_path=tflite_model_path)
interpreter.allocate_tensors()

input_index = interpreter.get_input_details()[0]['index']
output_index = interpreter.get_output_details()[0]['index']

classes = [
    'curly',
    'straight'
]

def prepare_image(img, target_size):
    if img.mode != 'RGB':
        img = img.convert('RGB')
    img = img.resize(target_size, Image.NEAREST)
    return img


def predict(url):
    image = download_image(url)
    image = prepare_image(image, target_size)
    
    x = np.array(image, dtype='float32')
    X = np.expand_dims(x, axis=0) 
    X = X / 255.0
    
    interpreter.set_tensor(input_index, X)
    interpreter.invoke()
    preds = interpreter.get_tensor(output_index)

    float_predictions = preds[0].tolist()

    return dict(zip(classes, float_predictions))


def lambda_handler(event, context):
    url = event['url']
    result = predict(url)
    return result
    

Overwriting lambda_function.py


## Docker 

For the next two questions, we'll use a Docker image that we already 
prepared. This is the Dockerfile that we used for creating the image:

```docker
FROM public.ecr.aws/lambda/python:3.10

COPY model_2024_hairstyle_v2.tflite .

RUN pip install numpy==1.23.1
```

Note that it uses Python 3.10. The latest models of TF Lite
do not support Python 3.12 yet, so we need to use 3.10. Also,
for this part, we will use TensorFlow 2.14.0. We have tested
it, and the models created with 2.17 could be served with 2.14.0.

For that image, we also needed to use an older version of numpy
(1.23.1)

The docker image is published to [`agrigorev/model-2024-hairstyle:v3`](https://hub.docker.com/r/agrigorev/model-2024-hairstyle/tags).

A few notes:

* The image already contains a model and it's not the same model
  as the one we used for questions 1-4.
* The wheel for this combination that you'll need to use in your Docker image is https://github.com/alexeygrigorev/tflite-aws-lambda/raw/main/tflite/tflite_runtime-2.14.0-cp310-cp310-linux_x86_64.whl


In [36]:
%%writefile Dockerfile
FROM agrigorev/model-2024-hairstyle:v3

RUN pip install --no-deps https://github.com/alexeygrigorev/tflite-aws-lambda/raw/main/tflite/tflite_runtime-2.14.0-cp310-cp310-linux_x86_64.whl
RUN pip install keras-image-helper

COPY lambda_function.py .

CMD [ "lambda_function.lambda_handler" ]

Overwriting Dockerfile


## Question 5

Download the base image `agrigorev/model-2024-hairstyle:v3`. You can do it with [`docker pull`](https://docs.docker.com/engine/reference/commandline/pull/).

So what's the size of this base image?

* 182 Mb
* 382 Mb
* 582 Mb
* 782 Mb

You can get this information when running `docker images` - it'll be in the "SIZE" column.


In [100]:
!docker pull agrigorev/model-2024-hairstyle:v3

v3: Pulling from agrigorev/model-2024-hairstyle

68a79b8a: Pulling fs layer 
124cce46: Pulling fs layer 
8b038848: Pulling fs layer 
7e7c1be9: Pulling fs layer 
8c0b7487: Pulling fs layer 
0580071d: Pulling fs layer 
e35356a0: Pulling fs layer 
Digest: sha256:6d0aa19127467401fd439c76e662ceba2bd21a5713f174d4f89b354a26939ea6[8A
Status: Downloaded newer image for agrigorev/model-2024-hairstyle:v3
docker.io/agrigorev/model-2024-hairstyle:v3


In [37]:
!docker image ls | grep agrigorev

agrigorev/model-2024-hairstyle                  v3             607850a2138e   5 days ago       782MB


So what's the size of this base image?

* 182 Mb
* 382 Mb
* 582 Mb
* **782 Mb**

## Question 6

Now let's extend this docker image, install all the required libraries
and add the code for lambda.

You don't need to include the model in the image. It's already included. 
The name of the file with the model is `model_2024_hairstyle_v2.tflite` and it's 
in the current workdir in the image (see the Dockerfile above for the 
reference). 
The provided model requires the same preprocessing for images regarding target size and rescaling the value range than used in homework 8.

Now run the container locally.

Score this image: https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg

What's the output from the model?

* 0.229
* 0.429
* 0.629
* 0.829

In [38]:
!docker build -t emmuzoo-2024-hairstyle .

[+] Building 0.0s (0/1)                                          docker:default
[+] Building 0.1s (8/8)                                          docker:default
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 328B                                       0.0s
 => [internal] load metadata for docker.io/agrigorev/model-2024-hairstyle  0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [1/4] FROM docker.io/agrigorev/model-2024-hairstyle:v3                 0.0s
 => [internal] load build context                                          0.0s
 => => transferring context: 1.58kB                                        0.0s
 => CACHED [2/4] RUN pip install --no-deps https://github.com/alexeygrigo  0.0s
 => CACHED [3/4] RUN pip install keras-image-helper                        0.0s
 => [4/4] COPY lambda_function.py .     

In [39]:
!docker images | grep emmuzoo-2024-hairstyle

emmuzoo-2024-hairstyle                          latest         6065fa26fea3   27 seconds ago   810MB


In [48]:
!docker run --rm -d -p 9000:8080 --name app emmuzoo-2024-hairstyle

e417d7257226044f649c7550d9fb33ad76f44e855ee8f3c46b6d43e86579933d


In [49]:
!docker ps -a | grep app

e417d7257226   emmuzoo-2024-hairstyle   "/lambda-entrypoint.…"   2 seconds ago   Up Less than a second   0.0.0.0:9000->8080/tcp, [::]:9000->8080/tcp   app


In [50]:
url = 'http://localhost:9000/2015-03-31/functions/function/invocations'

image_url = 'https://habrastorage.org/webt/yf/_d/ok/yf_dokzqy3vcritme8ggnzqlvwa.jpeg'

data = {'url': image_url}

result = requests.post(url, json=data)

In [51]:
print(result)

<Response [200]>


In [52]:
print(result.json())

{'curly': 0.42985352873802185}


In [53]:
!docker stop app
!docker ps -a

app
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES


What's the output from the model?

* 0.229
* **0.429**
* 0.629
* 0.829

## Publishing it to AWS

Now you can deploy your model to AWS!

* Publish your image to ECR
* Create a lambda function in AWS, use the ECR image
* Give it more RAM and increase the timeout 
* Test it
* Expose the lambda function using API Gateway

This is optional and not graded.

## Publishing to Docker hub


Just for the reference, this is how we published our image to Docker hub:

```bash
docker build -t model-2024-hairstyle -f homework.dockerfile .
docker tag model-2024-hairstyle:latest agrigorev/model-2024-hairstyle:v3
docker push agrigorev/model-2024-hairstyle:v3
```

(You don't need to execute this code)



## Submit the results


* Submit your results here: https://courses.datatalks.club/ml-zoomcamp-2024/homework/hw09
* If your answer doesn't match options exactly, select the closest one
